service/controller/resource/nodepool/create_deployment_uninitialized.go

```diff
@@ -133,6 +139,13 @@ func (r *Resource) deploymentUninitializedTransition(ctx context.Context, obj in
 	if deploymentNeedsToBeSubmitted {
 		r.Logger.LogCtx(ctx, "level", "debug", "message", "template or parameters changed")
 
+		if thisIsUpgrade && !conditions.IsUpgradingTrue(machinePool) {
```

Comment:
As I wrote in the NP upgrade PR's comment, I'd propose to keep the reconciliation logic straightforward and let the handler here only consider desired vs. present state, if possible.

My current thinking goes that upgrade logic should be possible to keep at higher level while provider specific implementation then ensures that running infrastructure matches to desired state that CR specification represents.

That way it should be hopefully easier to reason the overall logic of reconciliation. WDYT?

Reply:
> My current thinking goes that upgrade logic should be possible to keep at higher level while provider specific implementation then ensures that running infrastructure matches to desired state that CR specification represents.
> That way it should be hopefully easier to reason the overall logic of reconciliation. WDYT?

Yeah, that would be easier to reason about definitely, but I believe that we would have to change current `nodepool` handler implementation then.

Correct me if I'm wrong, this is my understanding of the current upgrade process on an example (for `nodepool` handler):
- we have a cluster with 5 node pools
- upgrade is triggered from `v20.0.0` to `v21.0.0`
- new azure-operator version starts to reconcile the cluster
- for every `AzureMachinePool`, `nodepool` handler compares current Azure ARM deployment/template to the desired ARM deployment/template
    - it finds the difference for at least azure-operator version
    - it submits a new VMSS deployment
    - it does the above deployment checks/submitting for all `AzureMachinePool` CRs
    - so effectively all node pools are upgraded in parallel (so rolling, draining and the whole shebang), which could disrupt customer's workload, and this is something that we wanted to avoid with sequential node pool upgrades (one node pool upgrade at a time), if I understood correctly how we wanted to have NP to NP upgrade.

So my understanding is that we would have to change current `nodepool` handler implementation, if we want to upgrade one node pool at a time.

Reply:
> it does the above deployment checks/submitting for all AzureMachinePool CRs
> so effectively all node pools are upgraded in parallel

No, the trick is that operator version label in `MachinePool` / `AzureMachinePool` CR guards this. As long as the labels are not updated, the old operator is still reconciling those CRs. Only once the labels are updated, the new operator picks up with them.

Reply:
If we can achieve to handle upgrades on a higher non-provider level (so I guess on CAPI level), that would really be more understandable.

I'm just not sure if that's possible in every scenario, as we sometimes have very specific upgrade steps to do, especially when upgrading/migrating to a new release where we have bigger changes, so it can be difficult to have a "dumb" handler that just "simply" ensures desired state that is finds in the CR.